File: app/BreedingRoller/old_python/Traits_Roller.py
import random

'''
Drakovai Type Calculator
1% chance to be a random dragon of higher rarity
5% chance to be a random dragon of either parent’s rarity
47/47 given parents are same rarity 
'''

'''Drakovai_rarity = {"wyvern":"common","drake":"common","ampithere":"common",
                   "true dragon":"uncommon","wyrm":"uncommon",
                   "lung":"rare",
                   "leviathan":"legendary"}

rarity_score = {"common":1,"uncommon":2,"rare":3,"legendary":4}'''

Drakovai_rarity = {"wyvern":"common","drake":"common",
                   "true dragon":"uncommon"}

rarity_score = {"common":1,"uncommon":2}

#Drakovai_rarity = {"wyvern":"common"}
#rarity_score = {"common":1}

    
def highest_rarity():
    high = max(rarity_score.values())
    type = [key for key, value in rarity_score.items() if value == high]
    return type[0]
    
def calculate_type(type1,type2):
    rarity1 = Drakovai_rarity[type1]
    rarity2 = Drakovai_rarity[type2]
    
    result = random.randint(1,100)

    if rarity_score[rarity1] - rarity_score[rarity2] > 0:
        high = [rarity1,type1]
        low = [rarity2,type2]
    else:
        high = [rarity2,type2]
        low = [rarity1,type1]

    match result:
        case 1 if high[0] == highest_rarity():
            # Case 1: 1% chance to be a random dragon of higher rarity (but there is no higher rarity)
            types = {key: value for key, value in Drakovai_rarity.items() if value == high[0]}
            child_type = random.choice(list(types.keys()))
        case 1:
            # Case 1: 1% chance to be a random dragon of higher rarity
            rarities = {key: value for key, value in rarity_score.items() if value == rarity_score[high[0]]+1}
            types = {key: value for key, value in Drakovai_rarity.items() if value in list(rarities.keys())}
            child_type = random.choice(list(types.keys()))
        case 2 | 3 | 4 | 5 | 6:
            # Case 2: 5% chance to be a random dragon of either parent's rarity
            types = {key: value for key, value in Drakovai_rarity.items() if value == high[0] or value == low[0]}
            child_type = random.choice(list(types.keys()))
        case _:
            # Default: inherit one of the parent's types
            score = rarity_score[high[0]] - rarity_score[low[0]]

            match score:
                case 0: # 50% chance of inheriting parent 1
                    child_type = random.choice([low[1], high[1]])
                case 1: # 66% chance of inheriting less rare type 
                    child_type = random.choice([low[1],low[1],high[1]])
                case 2: # 75% chance of inheriting less rare type 
                    child_type = random.choice([low[1],low[1],low[1],high[1]])
                case _: # 83% chance of inheriting less rare type 
                    child_type = random.choice([low[1],low[1],low[1],low[1],low[1],high[1]])

    return child_type


'''
Crossover Traits: If the baby dragon is one of the two parents types, there is a chance it can get one of the parent’s traits
90% chance of nothing
5% of one trait
3% of two traits
2% of three traits
'''

Drakovai_traits = {"wyvern":{"body variant":["scaled","feathered"],"head":["crest","none"],"mouth":["basic","primitive","beak"]},
                   "drake":{"body variant":["scaled","armored"],"head":["frill","none"],"mouth":["basic","primitive","tusk"]},
                   "true dragon":{"body variant":["scaled","armored"],"placeholder":["temp"],"mouth":["basic","primitive","fangs"]}}

def calculate_variantTraits(parent1, parent2, child):

    child_traits = {}

    if child != parent1 and child != parent2:
        traits = list(Drakovai_traits[child].keys())
        for trait in traits:
            choice = random.choice(Drakovai_traits[child][trait])
            child_traits[trait] = choice

    elif child == parent1 and child == parent2:
        traits = list(Drakovai_traits[child].keys())
        for trait in traits:
            choice = random.choice(Drakovai_traits[child][trait])
            child_traits[trait] = choice

    else:
        result = random.randint(1,100)

        if child == parent1:
            different = parent2
        else:
            different = parent1

        match result:
            case 1 | 2:
                traits = list(Drakovai_traits[different].keys())

                if len(traits) == 3:
                    for trait in traits:
                        choice = random.choice(Drakovai_traits[different][trait])
                        child_traits[trait] = choice
                    
                    traits = list(Drakovai_traits[child].keys())
                    for trait in traits:
                        if trait not in list(child_traits.keys()):
                            choice = random.choice(Drakovai_traits[child][trait])
                            child_traits[trait] = choice
                
                else:
                    for i in range(3):
                        trait = random.choice(traits)
                        while trait in child_traits:
                            trait = random.choice(traits)

                        choice = random.choice(Drakovai_traits[different][trait])
                        child_traits[trait] = choice

                    traits = list(Drakovai_traits[child].keys())

                    for trait in traits:
                        if trait not in list(child_traits.keys()):
                            choice = random.choice(Drakovai_traits[child][trait])
                            child_traits[trait] = choice
            
            case 3 | 4 | 5:
                traits = list(Drakovai_traits[different])
                for i in range(2):
                    trait = random.choice(traits)
                    while trait in child_traits:
                        trait = random.choice(traits)

                    choice = random.choice(Drakovai_traits[different][trait])
                    child_traits[trait] = choice

                traits = list(Drakovai_traits[child].keys())

                for trait in traits:
                    if trait not in list(child_traits.keys()):
                        choice = random.choice(Drakovai_traits[child][trait])
                        child_traits[trait] = choice

            case 6 | 7 | 8 | 9 | 10:
                traits = list(Drakovai_traits[different].keys())
                trait = random.choice(traits)

                choice = random.choice(Drakovai_traits[different][trait])
                child_traits[trait] = choice

                traits = list(Drakovai_traits[child].keys())

                for trait in traits:
                    if trait not in list(child_traits.keys()):
                        choice = random.choice(Drakovai_traits[child][trait])
                        child_traits[trait] = choice

            case _:
                traits = list(Drakovai_traits[child].keys())

                for trait in traits:
                    choice = random.choice(Drakovai_traits[child][trait])
                    child_traits[trait] = choice

    child_traits = sorted(child_traits.items())
    return child_traits



'''
Horn Location + Range
ALL Dragons
Cape (Head+Spine)
Always small horn range
Other locations dragon type specific
Neck
Thorax
Legs (includes shoulder +hip)
Tail

Horn Ranges
None: 0
Small: 1-5
Medium: 6-10
Large: 11+
'''

File: app/BreedingRoller/TraitsRoller.php
<?php

namespace App\BreedingRoller;


use Illuminate\Support\Facades\Log;

class TraitsRoller
{
    private static $Drakovai_rarity = [
        "wyvern" => "common",
        "drake" => "common",
        "true dragon" => "uncommon",
    ];

    private static $rarity_score = [
        "common" => 1,
        "uncommon" => 2,
    ];

    private static $Drakovai_traits = [
        "wyvern" => [
            "body variant" => ["scaled", "feathered"],
            "head" => ["crest", "none"],
            "mouth" => ["basic", "primitive", "beak"],
        ],
        "drake" => [
            "body variant" => ["scaled", "armored"],
            "head" => ["frill", "none"],
            "mouth" => ["basic", "primitive", "tusk"],
        ],
        "true dragon" => [
            "body variant" => ["scaled", "armored"],
            "placeholder" => ["temp"],
            "mouth" => ["basic", "primitive", "fangs"],
        ],
    ];

    public static function highestRarity()
    {
        $high = array_search(max(self::Drakovai_rarity), self::Drakovai_rarity);
        return $high;
    }

    public static function calculateType(string $type1, string $type2): string
    {
        $rarity1 = self::$Drakovai_rarity[$type1];
        $rarity2 = self::$Drakovai_rarity[$type2];
    
        $result = random_int(1, 100);
    
        if (self::$rarity_score[$rarity1] > self::$rarity_score[$rarity2]) {
            $high = [$rarity1, $type1];
            $low = [$rarity2, $type2];
        } else {
            $high = [$rarity2, $type2];
            $low = [$rarity1, $type1];
        }
    
        if ($result == 1) { // Case 1: 1% chance to be a random dragon of higher rarity (but there is no higher rarity)
            if ($high[0] === self::highestRarity()) { 
                $types = array_keys(array_filter(self::$Drakovai_rarity, fn($value) => $value === $high[0]));
                return $types[array_rand($types)];
            } else {
                $rarities = array_keys(array_filter(self::$rarity_score, fn($value) => $value === self::$rarity_score[$high[0]] + 1));
                $types = array_keys(array_filter(self::$Drakovai_rarity, fn($value) => in_array($value, $rarities)));
                return $types[array_rand($types)];
            }
        } elseif ($result <= 6) { // Case 2: 5% chance to be a random dragon of either parent's rarity
            $types = array_keys(array_filter(self::$Drakovai_rarity, fn($value) => $value === $high[0] || $value === $low[0]));
            return $types[array_rand($types)];
        } else {// Default: inherit one of the parent's types
            $score = self::$rarity_score[$high[0]] - self::$rarity_score[$low[0]];
    
            switch ($score) {
                case 0:
                    $possibleTypes = [$low[1], $high[1]];
                    break;
                case 1:
                    $possibleTypes = [$low[1], $low[1], $high[1]];
                    break;
                case 2:
                    $possibleTypes = [$low[1], $low[1], $low[1], $high[1]];
                    break;
                default:
                    $possibleTypes = [$low[1], $low[1], $low[1], $low[1], $low[1], $high[1]];
                    break;
            }
    
            return $possibleTypes[array_rand($possibleTypes)];
        }
    }

    public static function calculateVariantTraits(string $parent1, string $parent2, string $child): array
    {
        $child_traits = [];

        if ($child !== $parent1 && $child !== $parent2) {
            foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                $child_traits[$trait] = $options[array_rand($options)];
            }
        } elseif ($child === $parent1 && $child === $parent2) {
            foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                $child_traits[$trait] = $options[array_rand($options)];
            }
        } else {
            $result = random_int(1, 100);

            if ($child === $parent1) {
                $different = $parent2;
            } else {
                $different = $parent1;
            }

            switch ($result) {
                case 1:
                case 2:
                    $traits = array_keys(self::$Drakovai_traits[$different]);
                    if (count($traits) === 3) {
                        foreach ($traits as $trait) {
                            $child_traits[$trait] = self::$Drakovai_traits[$different][$trait][array_rand(self::$Drakovai_traits[$different][$trait])];
                        }
                        foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                            if (!isset($child_traits[$trait])) {
                                $child_traits[$trait] = $options[array_rand($options)];
                            }
                        }
                    } else {
                        for ($i = 0; $i < 3; $i++) {
                            $trait = $traits[array_rand($traits)];
                            while (isset($child_traits[$trait])) {
                                $trait = $traits[array_rand($traits)];
                            }
                            $child_traits[$trait] = self::$Drakovai_traits[$different][$trait][array_rand(self::$Drakovai_traits[$different][$trait])];
                        }
                        foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                            if (!isset($child_traits[$trait])) {
                                $child_traits[$trait] = $options[array_rand($options)];
                            }
                        }
                    }
                    break;
                case 3:
                case 4:
                case 5:
                    $traits = array_keys(self::$Drakovai_traits[$different]);
                    for ($i = 0; $i < 2; $i++) {
                        $trait = $traits[array_rand($traits)];
                        while (isset($child_traits[$trait])) {
                            $trait = $traits[array_rand($traits)];
                        }
                        $child_traits[$trait] = self::$Drakovai_traits[$different][$trait][array_rand(self::$Drakovai_traits[$different][$trait])];
                    }
                    foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                        if (!isset($child_traits[$trait])) {
                            $child_traits[$trait] = $options[array_rand($options)];
                        }
                    }
                    break;
                case 6:
                case 7:
                case 8:
                case 9:
                case 10:
                    $traits = array_keys(self::$Drakovai_traits[$different]);
                    $trait = $traits[array_rand($traits)];
                    $child_traits[$trait] = self::$Drakovai_traits[$different][$trait][array_rand(self::$Drakovai_traits[$different][$trait])];
                    foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                        if (!isset($child_traits[$trait])) {
                            $child_traits[$trait] = $options[array_rand($options)];
                        }
                    }
                    break;
                default:
                    foreach (self::$Drakovai_traits[$child] as $trait => $options) {
                        $child_traits[$trait] = $options[array_rand($options)];
                    }
            }
        }

        return $child_traits;
    }
}